feat(timeline-ui): [HUDI-9315] Add Hudi Timeline UI

voonhous opened the pull request:
### Describe the issue this Pull Request addresses

Visualizing the timeline via console and finder/explorer is often challenging as files are often sorted via naming convention. With completion time thrown into completed instants, it makes grouping of instants even harder when looking at it via console. 

Hence, a timeline visualizer can oftentimes make one able to understand the timeline better, especially for concurrent/long writes.

The timeline-ui page is accessible via `http://timeline_server__host:timeline_server__port/ui` after starting the timeline service.


### Summary and Changelog

Attached below is an example of how it looks like for now, functionalities are there, but it is a little rough:

#### Main page

The main page allow users to input the the table paths and the timeline will be visualised.

<img width="1025" height="335" alt="image" src="https://github.com/user-attachments/assets/25370c58-b037-4a31-9fbc-c8fc5430944a" />


#### Hovering over instants

Hovering over the instant bars will show you additional details like duration an action took.

<img width="4092" height="1386" alt="image" src="https://github.com/user-attachments/assets/96b4abb2-b708-4603-a9b9-9b6d4be4ce7e" />


#### Selecting an instant

Since commit details are saved to the filesystem as avro binaries, the commit information are not easily readable. Selecting the commit/deltacommit will return the details in json format.

<img width="2052" height="1155" alt="image" src="https://github.com/user-attachments/assets/495082af-b6fa-4bc1-87fa-5d943b2c3d5f" />


#### Clean range

Upon selection of a `clean` commit/plan that uses `KEEP_LATEST_COMMITS`, clean range will be displayed.

<img width="2043" height="746" alt="image" src="https://github.com/user-attachments/assets/20c06785-b38d-4c8e-a0cf-370d7320af37" />


#### Configs page

Table configurations are displayed in the `Table Config` tab:

<img width="2041" height="813" alt="image" src="https://github.com/user-attachments/assets/f607f0ff-751e-4647-b8c6-b66542a90212" />

#### Schema History

Schema related changelogs are also displayed (only when schema operations are done using `hoodie.schema.on.read.enable=true` under the `Shcema History` tab.

<img width="2042" height="1177" alt="image" src="https://github.com/user-attachments/assets/2281da8d-c972-429c-8a06-a2d3a1c82bd1" />


The schema changelog can also be expanded to see what column changed at which schemacommit.

<img width="2040" height="534" alt="image" src="https://github.com/user-attachments/assets/e05cd10e-3905-4ced-bc12-3a47512f8fba" />


### Impact

_Describe any public API or user-facing feature change or any performance impact._

Added 4 API endpoints on the `/v2/hoodie/view` path:
1. `timeline/instants/all?basepath={basepath}` - all instants (v2 format)
2. `timeline/instant?basepath={basepath}&instant={instant}&instantaction={action}&instantstate={state}` -  instant details
3. `table/config?basepath={basepath}` - table configuration
4. `table/schema/history?basepath={basepath}&limit={limit}` - schema evolution history

The UI is served at `/ui` (previously redirected to `/index.html`).

### Risk Level

None

### Documentation Update

We do not have any documentations for timeline service, hence, i suppose no documentation changes are required.

### Contributor's checklist

- [x] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [x] Change Logs and Impact were stated clearly
- [x] Adequate tests were added if applicable
- [x] CI passed

danny0405 commented:
@voonhous Nice feature, do you think it deserves a RFC then?

voonhous commented:
Sure, will create a new RFC for this.

I'll make a call out to **RFC-05** so that users will not get confused with it then:
https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=130027233

yihua commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/RequestHandler.java:
```diff
@@ -240,6 +252,30 @@ private void registerTimelineAPI() {
+
+    app.get(RemoteHoodieTableFileSystemView.INSTANT_DETAILS_URL, new ViewHandler(ctx -> {
+      metricsRegistry.add("INSTANT_DETAILS", 1);
+      Object instantDetailsObj = instantHandler.getInstantDetails(getBasePathParam(ctx),
```
🤖 nit: could you rename `instantDetailsObj` to `instantDetails`? The "Obj" suffix is non-standard in Java naming — the variable type is already declared as Object, so the suffix doesn't add clarity.

<sub><i>- Generated by an AI agent and may contain mistakes. Please verify any suggestions before applying.</i></sub>

voonhous replied:
Done, renamed `instantDetailsObj` to `instantDetails`.

yihua reviewed (commented):
🤖 *This review was generated by an AI agent and may contain mistakes. Please verify any suggestions before applying.*

**Style & Readability Review** — Code is generally clean and readable. One naming convention suggestion: the variable suffix "Obj" is non-standard in Java.

yihua commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/handlers/TimelineHandler.java:
```diff
@@ -46,4 +61,151 @@ public List<InstantDTO> getLastInstant(String basePath) {
+          HoodieInstant.State.valueOf(state), action, requestedTime,
+          InstantComparatorV1.REQUESTED_TIME_BASED_COMPARATOR);
+
+      Object result;
```
🤖 This hardcodes `InstantComparatorV1` regardless of the table's actual timeline layout version. If the table uses timeline layout V2 (Hudi 1.x default), the created instant will have a V1 comparator while the timeline's instants use V2, which could cause lookup failures or inconsistent ordering (V2 adds CLUSTERING action mapping that V1 lacks). Could you use the timeline's own `InstantGenerator` to create the instant instead?

<sub><i>- Generated by an AI agent and may contain mistakes. Please verify any suggestions before applying.</i></sub>

voonhous replied:
Good catch. `getInstantDetails` now derives the `InstantGenerator` from the timeline's own layout version (`TimelineLayout.fromVersion(timeline.getTimelineLayoutVersion()).getInstantGenerator()`) and uses it to create the instant, instead of hardcoding the V1 comparator. So a V2-layout table gets a V2 instant with the correct comparator and action mapping (including CLUSTERING). Removed the now-unused `InstantComparatorV1` import.

yihua commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/handlers/TimelineHandler.java:
```diff
@@ -46,4 +61,151 @@ public List<InstantDTO> getLastInstant(String basePath) {
+          result = requestedInstant.isCompleted()
+              ? hoodieTimeline.readCommitMetadata(requestedInstant)
+              : hoodieTimeline.readIndexPlan(requestedInstant);
+          break;
```
🤖 This catches all exceptions (including NPE, IllegalArgumentException, ClassCastException, etc.) and silently returns `null`, which the caller will serialize as JSON `null` with a 200 OK status. The client has no way to distinguish "instant not found" from "server error". I'd suggest at minimum logging a warning here, and ideally returning a proper error response with an appropriate status code.

<sub><i>- Generated by an AI agent and may contain mistakes. Please verify any suggestions before applying.</i></sub>

voonhous replied:
Good catch. `getInstantDetails` no longer collapses failures into a `null`/200 response:
- A malformed `state` or an unsupported `action` now returns 400 (`BadRequestResponse`).
- Any other read/deserialization failure is logged at WARN and surfaced as a 500 (`HoodieException`), matching the RFC's error-handling spec.

The caller can now tell a bad request from a server error. Also normalized this file's import order, which was failing checkstyle.

yihua commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/RequestHandler.java:
```diff
@@ -240,6 +252,30 @@ private void registerTimelineAPI() {
+    }, false));
+
+    app.get("/v2/hoodie/view/table/schema/history", new ViewHandler(ctx -> {
+      metricsRegistry.add("SCHEMA_HISTORY", 1);
```
🤖 The `Integer.parseInt` here will throw an uncaught `NumberFormatException` if a non-numeric `limit` value is provided. Consider wrapping this in a try-catch or using a safer parsing approach with a fallback to the default 200.

<sub><i>- Generated by an AI agent and may contain mistakes. Please verify any suggestions before applying.</i></sub>

voonhous replied:
Agreed. The `limit` param is now parsed safely and bounded: a non-numeric value returns a 400 (`BadRequestResponse`) instead of a raw `NumberFormatException`/500, and values outside 1..1000 are rejected with a 400. Still defaults to 200 when omitted.

yihua commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/RequestHandler.java:
```diff
@@ -240,6 +252,30 @@ private void registerTimelineAPI() {
       writeValueAsString(ctx, dto);
     }, false));
+
+    app.get(RemoteHoodieTableFileSystemView.TIMELINE_V2_URL, new ViewHandler(ctx -> {
```
🤖 These v2 API endpoints (timeline, instant details, table config, schema history) are registered unconditionally in `registerTimelineAPI()`, but the `enableUi` flag only gates the static file serving and the UiHandler. Should these v2 endpoints also be gated behind `enableUi` to avoid exposing table metadata when the UI feature is not enabled?

<sub><i>- Generated by an AI agent and may contain mistakes. Please verify any suggestions before applying.</i></sub>

voonhous replied:
Agreed, and this matches the RFC (v2 registration gated behind `--enable-ui`). Moved the four v2 routes (timeline, instant details, table config, schema history) into a new `registerTimelineV2API()` that is only called when `enableUi` is set, alongside the static files and `UiHandler`. The v1 endpoints (`LAST_INSTANT`, `TIMELINE`) stay unconditional. These v2 URLs are only consumed by the UI frontend, so nothing else is affected.

yihua commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/handlers/TimelineHandler.java:
```diff
@@ -46,4 +61,151 @@ public List<InstantDTO> getLastInstant(String basePath) {
+      // restrictions on Avro's internal Schema classes. Convert to plain Maps
+      // using JsonUtils which accesses fields directly, bypassing getSchema().
+      if (result instanceof SpecificRecordBase) {
+        return JsonUtils.getObjectMapper().convertValue(result, Map.class);
```
🤖 Both `getTableConfig` and `getSchemaHistory` create a new `HoodieTableMetaClient` on every request. MetaClient construction involves filesystem operations to read hoodie.properties and the timeline. Have you considered caching this or reusing the metaClient available through the `viewManager`?

<sub><i>- Generated by an AI agent and may contain mistakes. Please verify any suggestions before applying.</i></sub>

voonhous replied:
Good point, and it lines up with the RFC. Added a per-basepath `ConcurrentHashMap<String, HoodieTableMetaClient>` cache in `TimelineHandler` with a `getOrCreateMetaClient(basePath)` helper. `getTableConfig` and `getSchemaHistory` now reuse the cached metaClient instead of rebuilding one per request, so the hoodie.properties/timeline load is paid once per basepath. (`getInstantDetails` reads through the already-cached `FileSystemView` from `FileSystemViewManager`, so it does not build a metaClient at all.) Timeline freshness is unaffected since `getSchemaHistory` reloads via `getActiveTimeline()` on each call.

yihua commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/TimelineService.java:
```diff
@@ -250,7 +266,7 @@ public void run() throws IOException {
     HoodieLocalEngineContext localEngineContext = new HoodieLocalEngineContext(conf);
     // Just use defaults for now
-    HoodieMetadataConfig metadataConfig = HoodieMetadataConfig.newBuilder().build();
+    HoodieMetadataConfig metadataConfig = HoodieMetadataConfig.newBuilder().enable(false).build();
```
🤖 This changes the default metadata config for the standalone `TimelineService` from enabled (default `true`) to explicitly disabled. While this might be intentional for the standalone service, this affects all `buildFileSystemViewManager` callers — not just the timeline UI. Was this change intended to be part of this PR, or is it a separate fix? If intentional, it could be worth a comment explaining why metadata is disabled here.

<sub><i>- Generated by an AI agent and may contain mistakes. Please verify any suggestions before applying.</i></sub>

voonhous replied:
Intentional. The standalone timeline server (which backs the Timeline UI) serves read-only timeline data for arbitrary table basepaths, so it should not require each table to have a metadata table present or bootstrapped; with MDT disabled it falls back to direct storage listing. Replaced the stale "Just use defaults for now" comment with this rationale.

Scope is limited to the standalone `buildFileSystemViewManager` path (only called from `TimelineService` startup). The embedded server is unaffected: it builds its own view manager from the writer's metadata config via `EmbeddedTimelineService.createViewManager()`.

yihua reviewed (commented):
🤖 *This review was generated by an AI agent and may contain mistakes. Please verify any suggestions before applying.*

Thanks for contributing! The timeline UI is a nice addition for debugging and understanding Hudi timelines. A few functional issues to address: the hardcoded `InstantComparatorV1` in `getInstantDetails` could cause lookup failures on V2-layout tables, the broad `catch (Exception)` silently swallows errors returning null with HTTP 200, and `Integer.parseInt` on the user-provided `limit` param can throw uncaught exceptions. See inline comments for details.

nsivabalan commented:
This is pretty amazing @voonhous 👏

hudi-agent reviewed (commented):
🤖 *This review was generated by an AI agent and may contain mistakes. Please verify any suggestions before applying.*

The new change in this update is a single pom.xml addition that excludes `**/resources/public/lib/**` from the Apache RAT license-header check, with a comment noting that the bundled third-party UI libraries (Bootstrap, renderjson, vis-timeline) ship with their own licenses. The change looks appropriate and well-scoped. The prior inline findings (hardcoded `InstantComparatorV1` in `getInstantDetails`, broad `catch (Exception)` returning null, unsafe `Integer.parseInt` on the `limit` param, default metadata config flip in `buildFileSystemViewManager`, ungated v2 API endpoints, and per-request `HoodieTableMetaClient` construction) do not appear to have been addressed in this update — please take another look at those when you get a chance. Please take a look at any inline comments, and this should be ready for a Hudi committer or PMC member to take it from here.

hudi-agent commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/RequestHandler.java:
```diff
@@ -171,6 +171,18 @@ private static String getMinInstantParam(Context ctx) {
+
+  private static String getInstantStateParam(Context ctx) {
+    return ctx.queryParamAsClass(RemoteHoodieTableFileSystemView.INSTANT_STATE_PARAM, String.class).getOrDefault("");
+  }
```
_⚠️ Potential issue_ | _🟠 Major_ | _⚡ Quick win_

**Reject incomplete instant-detail requests up front.**

Lines 175, 179, and 183 default missing params to `""`, so the new instant-details route can call `getInstantDetails(...)` with an incomplete identifier tuple. That should be a 400 at the edge instead of an empty-string lookup downstream.  
 

<details>
<summary>Suggested fix</summary>

```diff
 private static String getInstantParam(Context ctx) {
-  return ctx.queryParamAsClass(RemoteHoodieTableFileSystemView.INSTANT_PARAM, String.class).getOrDefault("");
+  return ctx.queryParamAsClass(RemoteHoodieTableFileSystemView.INSTANT_PARAM, String.class)
+      .getOrThrow(e -> new HoodieException("INSTANT_PARAM is invalid"));
 }
 
 private static String getInstantActionParam(Context ctx) {
-  return ctx.queryParamAsClass(RemoteHoodieTableFileSystemView.INSTANT_ACTION_PARAM, String.class).getOrDefault("");
+  return ctx.queryParamAsClass(RemoteHoodieTableFileSystemView.INSTANT_ACTION_PARAM, String.class)
+      .getOrThrow(e -> new HoodieException("INSTANT_ACTION_PARAM is invalid"));
 }
 
 private static String getInstantStateParam(Context ctx) {
-  return ctx.queryParamAsClass(RemoteHoodieTableFileSystemView.INSTANT_STATE_PARAM, String.class).getOrDefault("");
+  return ctx.queryParamAsClass(RemoteHoodieTableFileSystemView.INSTANT_STATE_PARAM, String.class)
+      .getOrThrow(e -> new HoodieException("INSTANT_STATE_PARAM is invalid"));
 }
```
</details>

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In
`@hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/RequestHandler.java`
around lines 174 - 184, The three helper methods getInstantParam,
getInstantActionParam, and getInstantStateParam currently return empty strings
for missing query params which allows getInstantDetails(...) to be called with
incomplete identifiers; change each helper to validate presence and reject
missing values immediately by returning a 400 error response (or throwing a
BadRequest/HttpResponseException handled by the router) instead of defaulting to
"", so callers always receive a non-empty value or the request is terminated
with 400. Locate and update getInstantParam, getInstantActionParam, and
getInstantStateParam to check ctx.queryParamAsClass(...).isPresent() (or
equivalent), set ctx.status(400) with a clear message (e.g., "missing required
instant parameter: <name>") or throw a BadRequest exception when absent,
ensuring downstream getInstantDetails(...) is never called with empty
identifiers.
```

</details>

<!-- fingerprinting:phantom:medusa:grasshopper:2ed098e2-9d4f-4382-b4df-b34fcb790079 -->

<!-- 4e71b3a2 -->

<!-- This is an auto-generated comment by CodeRabbit -->

— *CodeRabbit* ([original](https://github.com/hudi-agent/hudi/pull/24#discussion_r3176964600)) (source:comment#3176964600)

voonhous replied:
Good catch. Made these mandatory: `getInstantParam`, `getInstantActionParam`, and `getInstantStateParam` now `getOrThrow` a `BadRequestResponse` (400) instead of defaulting to `""`, so a missing identifier is rejected at the edge rather than reaching `getInstantDetails` with an empty value. These getters feed only the `/v2` instant-details route, so nothing else is affected.

hudi-agent commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/RequestHandler.java:
```diff
@@ -240,6 +252,30 @@ private void registerTimelineAPI() {
+    app.get("/v2/hoodie/view/table/schema/history", new ViewHandler(ctx -> {
+      metricsRegistry.add("SCHEMA_HISTORY", 1);
+      int limit = Integer.parseInt(ctx.queryParamAsClass("limit", String.class).getOrDefault("200"));
+      writeValueAsString(ctx, instantHandler.getSchemaHistory(getBasePathParam(ctx), limit));
```
_⚠️ Potential issue_ | _🟠 Major_ | _⚡ Quick win_

**Validate and bound `limit` before using it.**

Line 276 throws a raw `NumberFormatException` for malformed input, which turns a client mistake into a 500. It also lets zero/negative or arbitrarily large values through unchecked. Return a `BadRequestResponse` for invalid values and cap the upper bound.  
 

<details>
<summary>Suggested fix</summary>

```diff
     app.get("/v2/hoodie/view/table/schema/history", new ViewHandler(ctx -> {
       metricsRegistry.add("SCHEMA_HISTORY", 1);
-      int limit = Integer.parseInt(ctx.queryParamAsClass("limit", String.class).getOrDefault("200"));
+      String rawLimit = ctx.queryParamAsClass("limit", String.class).getOrDefault("200");
+      int limit;
+      try {
+        limit = Integer.parseInt(rawLimit);
+      } catch (NumberFormatException e) {
+        throw new BadRequestResponse("limit must be a positive integer");
+      }
+      if (limit <= 0 || limit > 1000) {
+        throw new BadRequestResponse("limit must be between 1 and 1000");
+      }
       writeValueAsString(ctx, instantHandler.getSchemaHistory(getBasePathParam(ctx), limit));
     }, false));
```
</details>

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In
`@hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/RequestHandler.java`
around lines 274 - 277, The code in RequestHandler's route lambda reads
ctx.queryParamAsClass(...) into an int without validation, causing
NumberFormatException and allowing zero/negative or huge limits; update the
ViewHandler lambda to parse and validate the "limit" param (using ctx.queryParam
or queryParamAsClass), catch NumberFormatException and throw a
BadRequestResponse with a clear message for malformed input, enforce a minimum
of 1 and cap to a sensible maximum (e.g., 1000) before calling
instantHandler.getSchemaHistory(getBasePathParam(ctx), limit), and ensure
writeValueAsString is only called with the validated/capped limit.
```

</details>

<!-- fingerprinting:phantom:medusa:grasshopper:2ed098e2-9d4f-4382-b4df-b34fcb790079 -->

<!-- 4e71b3a2 -->

<!-- This is an auto-generated comment by CodeRabbit -->

— *CodeRabbit* ([original](https://github.com/hudi-agent/hudi/pull/24#discussion_r3176964601)) (source:comment#3176964601)

voonhous replied:
Agreed. The `limit` param is now parsed safely and bounded: a non-numeric value returns a 400 (`BadRequestResponse`) instead of a raw `NumberFormatException`/500, and values outside 1..1000 are rejected with a 400. Still defaults to 200 when omitted.

hudi-agent commented on hudi-timeline-service/src/main/resources/public/index.html:
```diff
@@ -0,0 +1,335 @@
+  <!-- Bootstrap 5 JS: CDN with bundled fallback -->
+  <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.3.3/dist/js/bootstrap.bundle.min.js"
+          onerror="var s=document.createElement('script');s.src='/lib/bootstrap/bootstrap.bundle.min.js';document.head.appendChild(s);"></script>
+  <script src="/js/timeline.js"></script>
```
_⚠️ Potential issue_ | _🟡 Minor_ | _⚡ Quick win_

**Script fallback race condition risk.**

When the CDN fails and the `onerror` handler dynamically creates a fallback script, `timeline.js` (Line 333) may execute before the fallback scripts finish loading, causing `vis`, `renderjson`, or `bootstrap` to be undefined.

Consider using `defer` and a more robust fallback mechanism that ensures dependencies load before `timeline.js`:

<details>
<summary>Suggested approach</summary>

```diff
-  <script src="https://unpkg.com/vis-timeline@7.7.3/standalone/umd/vis-timeline-graph2d.min.js"
-          onerror="var s=document.createElement('script');s.src='/lib/vis-timeline/vis-timeline-graph2d.min.js';document.head.appendChild(s);"></script>
+  <script src="https://unpkg.com/vis-timeline@7.7.3/standalone/umd/vis-timeline-graph2d.min.js"
+          onerror="this.onerror=null;var s=document.createElement('script');s.src='/lib/vis-timeline/vis-timeline-graph2d.min.js';document.head.appendChild(s);"></script>
```

Additionally, `timeline.js` should defensively check that dependencies are loaded, or you could wrap `timeline.js` in a `DOMContentLoaded` handler that polls for the presence of `vis`, `renderjson`, and `bootstrap`.

</details>

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `@hudi-timeline-service/src/main/resources/public/index.html` around lines 325
- 333, The current approach may let /js/timeline.js run before fallback scripts
load when CDN fetches fail; update the fallback logic so timeline.js is only
appended/executed after all dependency scripts (vis, renderjson, bootstrap) are
confirmed loaded: modify the inline onerror handlers to create their fallback
<script> elements with onload callbacks (or return Promises) and then append
/js/timeline.js only once all three onload handlers have fired (or use
Promise.all to wait for them), and also add defensive guards inside timeline.js
that check for the global symbols vis, renderjson, and bootstrap before using
them.
```

</details>

<!-- fingerprinting:phantom:medusa:ocelot:9bea6e99-c2b2-493c-ab46-668b4812a792 -->

<!-- 4e71b3a2 -->

<!-- This is an auto-generated comment by CodeRabbit -->

— *CodeRabbit* ([original](https://github.com/hudi-agent/hudi/pull/24#discussion_r3176964602)) (source:comment#3176964602)

voonhous replied:
Fixed by removing the CDN entirely and serving the bundled assets directly, which also matches the RFC update (we dropped the CDN-first strategy in favor of bundled-only). All five references (vis-timeline JS/CSS, bootstrap JS/CSS, renderjson JS) now load from `/lib/` with no `onerror` fallback, so there are no external calls and no fallback race: scripts load in document order from the local server and `timeline.js` runs last, after its dependencies.

hudi-agent commented on pom.xml:
```diff
@@ -670,6 +670,8 @@
               <exclude>.github/**</exclude>
               <exclude>**/banner.txt</exclude>
+              <!-- bundled third-party UI libraries (Bootstrap, renderjson, vis-timeline) ship with their own licenses -->
+              <exclude>**/resources/public/lib/**</exclude>
```
_⚠️ Potential issue_ | _🟠 Major_

<details>
<summary>🧩 Analysis chain</summary>

🏁 Script executed:

```shell
#!/bin/bash
set -euo pipefail

echo "Vendored assets under public/lib:"
fd -HI . hudi-timeline-service/src/main/resources/public/lib

echo
echo "License-ish files shipped next to vendored assets:"
fd -HI '^(LICENSE|NOTICE|COPYING|README|package\.json)$' hudi-timeline-service/src/main/resources/public/lib || true

echo
echo "Mentions in repository redistribution metadata:"
rg -n "Bootstrap|renderjson|vis-timeline|visjs|David Caldwell" LICENSE NOTICE || true
```

Repository: hudi-agent/hudi

Length of output: 817

---



</details>

**Blanket RAT exclusion for `public/lib` creates a compliance gap.**

Lines 673-674 exclude all files under `resources/public/lib/`, allowing bundled assets to bypass Apache RAT checks without requiring redistribution metadata. The three vendored libraries (Bootstrap, renderjson, vis-timeline) currently shipped in this directory lack local LICENSE files and are not documented in the repository-level LICENSE or NOTICE. Instead of a wildcard exclusion, scope this to the specific vendored directories and ensure each library's licensing information is captured in the repository's redistribution metadata.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against the current code and only fix it if needed.

In `@pom.xml` around lines 673 - 674, The current blanket RAT exclusion
"<exclude>**/resources/public/lib/**</exclude>" creates a compliance gap;
replace it with targeted exclusions for the three vendored library directories
and add license metadata: stop excluding the whole resources/public/lib tree,
instead exclude only the specific directories for Bootstrap, renderjson, and
vis-timeline (reference the "<exclude>**/resources/public/lib/...</exclude>"
entries in pom.xml), add each library's LICENSE file into the corresponding
vendor subdirectory (e.g., resources/public/lib/bootstrap/LICENSE, etc.), and
update the repository-level redistribution metadata/NOTICE (and the pom's
license/notice configuration) to list these three libraries and their licenses
so RAT and downstream redistributions capture proper license info.
```

</details>

<!-- fingerprinting:phantom:medusa:grasshopper:2ed098e2-9d4f-4382-b4df-b34fcb790079 -->

<!-- 4e71b3a2 -->

<!-- This is an auto-generated comment by CodeRabbit -->

— *CodeRabbit* ([original](https://github.com/hudi-agent/hudi/pull/24#discussion_r3176964603)) (source:comment#3176964603)

voonhous replied:
Intentional, ignore.

hudi-agent reviewed (commented):
🤖 *This review was generated by an AI agent and may contain mistakes. Please verify any suggestions before applying.*

**CodeRabbit Walkthrough:** This PR introduces a web-based timeline explorer UI for the Hudi Timeline Service. It adds v2 API endpoints to expose timeline and instant data as DTOs, implements backend handlers to serve timeline/config/schema information, and provides a complete frontend with Bootstrap/vis-timeline for interactive timeline visualization, filtering, and instant detail inspection.

**Sequence Diagram (CodeRabbit):**

```mermaid
sequenceDiagram
    participant Browser as Browser / UI
    participant TimelineService as Timeline Service
    participant FileSystemView as FileSystem View<br/>(Metadata Cache)
    participant HoodieTimeline as HoodieTimeline

    Browser->>TimelineService: GET /v2/hoodie/view/timeline/instants
    TimelineService->>FileSystemView: getTimeline()
    FileSystemView->>HoodieTimeline: getInstantsAsStream()
    HoodieTimeline-->>FileSystemView: stream of HoodieInstant
    FileSystemView-->>TimelineService: HoodieTimeline
    TimelineService->>TimelineService: TimelineDTO.fromTimeline()
    TimelineService-->>Browser: TimelineDTO (JSON)

    Browser->>Browser: vis.Timeline render<br/>with instants

    Browser->>TimelineService: GET /v2/hoodie/view/instants/details?instant=...&action=COMMIT
    TimelineService->>FileSystemView: getTimeline()
    FileSystemView->>HoodieTimeline: readCommitted() or readPending()
    HoodieTimeline-->>FileSystemView: SpecificRecordBase (metadata/plan)
    FileSystemView-->>TimelineService: Avro record
    TimelineService->>TimelineService: JsonUtils.convert to Map
    TimelineService-->>Browser: Detail JSON

    Browser->>Browser: Render detail card<br/>+ compute clean range<br/>if KEEP_LATEST_COMMITS
```

CodeRabbit: https://github.com/hudi-agent/hudi/pull/24#pullrequestreview-4215300877

hudi-agent commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/RequestHandler.java:
```diff
@@ -240,6 +252,30 @@ private void registerTimelineAPI() {
+      writeValueAsString(ctx, instantDetailsObj);
+    }, false));
+
+    app.get("/v2/hoodie/view/table/config", new ViewHandler(ctx -> {
```
🤖 nit: these two routes use inline string literals while every other route in this class references a named constant from `RemoteHoodieTableFileSystemView`. Could you add `TABLE_CONFIG_V2_URL` and `SCHEMA_HISTORY_V2_URL` constants there (line 274 is the same issue)?

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

voonhous replied:
Done. Added `TABLE_CONFIG_V2_URL` and `SCHEMA_HISTORY_V2_URL` to `RemoteHoodieTableFileSystemView` (next to the other v2 URL constants) and switched the two routes to reference them instead of inline string literals.

hudi-agent commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/handlers/TimelineHandler.java:
```diff
@@ -46,4 +61,151 @@ public List<InstantDTO> getLastInstant(String basePath) {
     return TimelineDTO.fromTimeline(viewManager.getFileSystemView(basePath).getTimeline());
   }
+
+  public org.apache.hudi.common.table.timeline.dto.v2.TimelineDTO getTimelineV2(String basePath) {
```
🤖 nit: the fully-qualified `org.apache.hudi.common.table.timeline.dto.v2.TimelineDTO` return type (and the same on line 66) is painful to read. Would it be possible to rename the v2 class — e.g. `TimelineDTOV2` — so it can be imported cleanly alongside the v1 `TimelineDTO`?

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

voonhous replied:
Done. Renamed the v2 `TimelineDTO` to `TimelineDTOV2` (file renamed via `git mv` to keep history). `RequestHandler` and `TimelineHandler` now import and use `TimelineDTOV2` instead of the fully-qualified name, so it reads cleanly alongside the v1 `TimelineDTO`. JSON wire format is unchanged (serialization is by `@JsonProperty` field names, not the class name). Left the v2 `InstantDTO` as-is since it is only referenced within its package and was not flagged, but happy to rename it for symmetry if you prefer.

hudi-agent commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/RequestHandler.java:
```diff
@@ -240,6 +252,30 @@ private void registerTimelineAPI() {
+
+    app.get(RemoteHoodieTableFileSystemView.INSTANT_DETAILS_URL, new ViewHandler(ctx -> {
+      metricsRegistry.add("INSTANT_DETAILS", 1);
+      Object instantDetailsObj = instantHandler.getInstantDetails(getBasePathParam(ctx),
```
🤖 nit: `instantDetailsObj` — the `Obj` suffix is redundant in Java since everything is an object. `instantDetails` or just `details` would be cleaner.

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

voonhous replied:
Done, renamed `instantDetailsObj` to `instantDetails`.

hudi-agent reviewed (commented):
🤖 *This review was generated by an AI agent and may contain mistakes. Please verify any suggestions before applying.*

Thanks for working on this! This PR adds a Timeline UI feature and supporting v2 endpoints (timeline, instant details, table config, schema history) to the timeline service. Prior review rounds have already flagged the key issues (parameter validation, error handling, script-fallback race, license exclusion). No new critical correctness issues flagged from this automated pass — a Hudi committer or PMC member can take it from here for a final review. A few naming and consistency suggestions below — two hard-coded URL strings that should be constants, a fully-qualified return type caused by a class name collision, and a minor variable naming nit.

cc @yihua

voonhous committed:
feat(timeline-ui): [HUDI-9315] Add Hudi Timeline UI

Add a read-only Timeline UI, served by hudi-timeline-service, for visualizing a
Hudi table's active timeline.

Backend:
- New v2 endpoints (timeline, instant details, table config, schema history) in
  TimelineHandler, registered via registerTimelineV2API() and gated behind the
  --enable-ui flag.
- Per-basepath HoodieTableMetaClient caching; layout-aware InstantGenerator when
  reading instant details; request-parameter validation (400 for missing or
  malformed params, bounded schema-history limit) and proper error responses
  instead of returning null.

Frontend:
- Static HTML/JS/CSS (vis-timeline, Bootstrap, renderjson) served from bundled
  assets only, with no external/CDN calls.

Integration:
- Embedded-mode startup and a Spark UI "Hudi Timeline" tab.
- Exclude bundled UI libraries from the apache-rat check.

hudi-agent commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/dto/v2/InstantDTO.java:
```diff
@@ -0,0 +1,63 @@
+  }
+
+  public static HoodieInstant toInstant(InstantDTO dto, InstantGenerator factory) {
+    if (null == dto) {
```
🤖 nit: could you rename `factory` to `instantGenerator` here? The same type is referred to as `instantGenerator` in `TimelineDTOV2.toTimeline` and in `TimelineHandler`, so calling it `factory` in this one spot breaks the naming convention and makes readers wonder whether it's a different object.

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

hudi-agent commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/RequestHandler.java:
```diff
@@ -242,6 +258,43 @@ private void registerTimelineAPI() {
+
+    app.get(RemoteHoodieTableFileSystemView.SCHEMA_HISTORY_V2_URL, new ViewHandler(ctx -> {
+      metricsRegistry.add("SCHEMA_HISTORY", 1);
+      int limit;
```
🤖 nit: the `limit` parsing and validation is done inline here, while every other query param in this class is extracted to a private helper (`getInstantParam`, `getBasePathParam`, etc.). It might be worth pulling this out to something like `getLimitParam(Context ctx)` to keep the pattern consistent and make the lambda body easier to scan.

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

voonhous replied:
Addressed

hudi-agent commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/handlers/TimelineHandler.java:
```diff
@@ -46,4 +78,149 @@ public List<InstantDTO> getLastInstant(String basePath) {
+
+    Map<String, Object> result = new HashMap<>();
+
+    // Get current schema
```
🤖 nit: `throws Exception` is broader than needed here — every checked exception in the method body is caught internally (either swallowed or rethrown as `HoodieException`/`BadRequestResponse`). Could you narrow this to `throws IOException` or remove the throws clause entirely? As-is, callers see `throws Exception` and have to decide what to catch unnecessarily.

<sub><i>- AI-generated; verify before applying. React 👍/👎 to flag quality.</i></sub>

hudi-agent reviewed (commented):
🤖 *This review was generated by an AI agent and may contain mistakes. Please verify any suggestions before applying.*

Thanks for iterating on this! At this point the timeline UI plumbing — v2 DTOs, the gated `/v2` routes, the new `/ui` static handler, instant-detail/table-config/schema-history endpoints, and the metaClient cache — looks consistent with how it's exercised by the bundled `timeline.js`. I traced the timeline source (`viewManager.getFileSystemView(basePath).getTimeline()` resolves to `visibleActiveTimeline` from `IncrementalTimelineSyncFileSystemView`, which is `filterCompletedAndCompactionInstants()` — so completed clean/rollback/restore/savepoint actions plus all compaction states are visible to the UI as expected), the completion-time lookup path in `getInstantFileName`, the `ConcurrentHashMap.computeIfAbsent` semantics for the metaClient cache, and the `Thread.currentThread().join()` shutdown pattern, and didn't surface anything new beyond what prior rounds already flagged and the author addressed. No correctness issues found. A few style/readability suggestions in the inline comments. Please take a look, and this should be ready for a Hudi committer or PMC member to take it from here. A few naming and simplification suggestions below.

cc @yihua

voonhous committed:
fix(timeline-ui): [HUDI-9315] Align Timeline UI with RFC-94 spec

Rework the Timeline UI implementation to match the merged RFC-94 spec:

- Move UI REST routes from /v2/hoodie/view/* to /ui/api/* and serve
  static assets under /ui/static; revert all changes to the FSV protocol
  class RemoteHoodieTableFileSystemView (the UI is not part of the FSV
  RPC contract)
- Replace dto.v2 with one-way dto.ui DTOs; UiInstantDTO carries a
  server-computed comparableAction
- Serve the timeline from the active timeline via a per-request
  metaClient (drops the stale metaClient cache and the FSV write
  timeline, which lacks clean/rollback/savepoint/restore/indexing and
  pending states)
- Resolve instant-detail requests against the active timeline instead of
  reconstructing instants from request params (path-traversal fix; 404
  for unknown instants) and route all instant content reads through a
  shared try-with-resources stream helper
- Schema history: non-throwing schema resolution, baseline/change entry
  typing, window {oldestInstantScanned, truncated}; limit clamps at 1000
- Fail fast when --enable-ui is set but UI assets are not bundled;
  exclude public/** from the five engine bundles so only
  hudi-timeline-server-bundle ships the vendored assets
- Add LICENSE stanzas for vis-timeline (MIT), Bootstrap (MIT) and
  renderjson (ISC); inject the root LICENSE into the server bundle jar;
  add scripts/release/validate_bundle_ui_assets.sh guarding the bundle
  asset exclusions
- Frontend: group rows by comparable action, Refresh controls on the
  Table Config and Schema History tabs, baseline/truncation rendering
- Add TestUiApi covering DTO mapping, instant details, path traversal,
  schema history, flag gating and static serving

voonhous committed:
fix(timeline-ui): [HUDI-9315] Close RFC-94 alignment gaps found in review

- validate_bundle_ui_assets.sh: the engine-bundle glob patterns never matched
  the scala-suffixed spark/utilities jar names (hudi-spark3.5-bundle_2.12-*,
  hudi-utilities-bundle_2.12-*), so the load-bearing negative assertion
  silently skipped both bundles; fix the patterns
- TimelineHandler: return 400 for an instantaction outside
  VALID_ACTIONS_IN_TIMELINE (RFC-94: a malformed state/action is a 400, only
  an absent instant is a 404)
- TestUiApi: pin the logcompaction -> deltacommit and clustering ->
  replacecommit comparable-action folds, cover the clean-plan instant detail,
  assert an empty timeline returns an empty instants list, and assert a
  malformed action returns 400
- timeline.js: surface the server's descriptive error body in alerts instead
  of statusText only (shared httpError helper, textContent-safe); always
  render the commit-schema history with its window truncation notice and
  baseline/change badges, rendering internal schema history as an additional
  labelled section instead of replacing them

voonhous committed:
fix(timeline-ui): [HUDI-9315] Fix instant-details reads for replacecommit, indexing and inflight instants

Three action/state combinations of /ui/api/timeline/instant returned 500:

- A completed replacecommit/clustering file is avro HoodieReplaceCommitMetadata
  on disk; reading it as avro HoodieCommitMetadata fails avro record-name
  resolution. Read the POJO HoodieReplaceCommitMetadata instead, mirroring
  TimelineUtils.getCommitMetadata. Apply the same fix to getSchemaHistory,
  which silently skipped schema changes delivered by insert_overwrite.
- A completed indexing instant stores HoodieIndexCommitMetadata, not
  HoodieCommitMetadata.
- The inflight file of a plan-carrying action is empty by design, so reading
  the plan from an INFLIGHT instant always failed; read the requested twin
  instead. Only timeline-resolved fields flow into the twin, preserving the
  path-traversal defense.

Also:
- Escape interpolated values in the vis-timeline tooltip, accept legacy
  14-digit timestamps in parseHudiTimestamp, and treat any all-zero request
  time as the init sentinel.
- Move the UI DTOs out of hudi-common into the timeline service, their only
  consumer.
- Note in the --enable-ui description that the timeline service has no
  authentication.

TestUiApi covers each previously-500 branch and the insert_overwrite schema
change.

voonhous commented:
@wombatu-kun Can you please help to review this? 

This is the first cut of the `hudi-ui`. After this lands, there's a bunch of cleanup that we need to do before developing on this further. 

First of which is to create an interface for table versions as each version may have different timeline specs.

voonhous committed:
refactor(timeline-ui): [HUDI-9315] extract schema-history limit parsing to a param helper

The limit query param was parsed and validated inline in the route lambda
while every other param in RequestHandler has a getXxxParam(Context) helper.
Extract getLimitParam with the same behavior: default 200, 400 on
non-numeric or non-positive values, clamp above 1000.

wombatu-kun commented on hudi-timeline-service/src/main/resources/public/js/timeline.js:
```diff
@@ -0,0 +1,1243 @@
+    var pattern = /^(\d{4})(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})(\d{3})$/;
+    var match = ts.match(pattern);
+    if (match) {
+      return match[1] + '-' + match[2] + '-' + match[3]
```
parseHudiTimestamp captures the milliseconds group in its 17-digit regex but omits it from the returned string, so item.start/item.end lose sub-second precision and computeStats' "if (diff > 0)" check silently drops every same-second commit from the average duration and collapses its range bar to zero width. Append the captured milliseconds group to the returned string so sub-second instants keep their real duration.

voonhous replied:
Good catch. Fixed -- `parseHudiTimestamp` now appends the captured milliseconds group, so same-second instants keep their sub-second durations in `computeStats`.

wombatu-kun commented on hudi-timeline-service/src/main/resources/public/js/timeline.js:
```diff
@@ -0,0 +1,1243 @@
+    });
+
+    newerFields.forEach(function (f) {
+      if (olderMap[f.name] && olderMap[f.name].type !== f.type) {
```
diffInternalSchemas compares field types with "!==" and renders them with String(), so any complex-typed field (record, array, map) is always reported as changed and shown as "[object Object] -> [object Object]", unlike diffSchemas which compares via JSON.stringify and formats via formatAvroType. Compare and render these types the same way diffSchemas does; follow-up, not a blocker.

voonhous replied:
Fixed in this PR anyway. `diffInternalSchemas` now compares via `JSON.stringify` and renders through a new `formatInternalType` (the internal-schema JSON uses `element` / `key` / `value` rather than Avro's `items` / `values`, so reusing `formatAvroType` as-is would print `array<undefined>`).

wombatu-kun commented on hudi-timeline-service/src/main/resources/public/js/timeline.js:
```diff
@@ -0,0 +1,1243 @@
+                if (prevJson && prevJson.earliestCommitToRetain) {
+                  var rangeStart = parseHudiTimestamp(prevJson.earliestCommitToRetain);
+                  var rangeEnd = parseHudiTimestamp(json.earliestCommitToRetain);
+                  allItems.add({
```
The clean-range background is removed synchronously in onSelect but re-added inside a nested async fetch, so selecting another instant or pressing Escape before that fetch resolves re-renders the previous instant's range against the new selection. Capture the current selection id before the fetch and add the range only if it still matches when the promise resolves; low-impact follow-up.

voonhous replied:
Fixed. The callback now checks `timeline.getSelection()[0]` against the fetched instant's id and bails if they no longer match -- covers re-selection, arrow-key nav and Escape (which clears the selection without going through `onSelect`).

wombatu-kun commented on scripts/release/validate_bundle_ui_assets.sh:
```diff
@@ -0,0 +1,152 @@
+# The Timeline UI assets (vis-timeline, Bootstrap, renderjson) must ship ONLY in
+# hudi-timeline-server-bundle. The five engine bundles exclude "public/**" so that
+# the bundled-library LICENSE obligations stay scoped to the server bundle. RFC-94
+# calls this negative assertion load-bearing: without it the exclusion can regress
```
This guard script is invoked by nothing - not validate_staged_release.sh, not bot.yml - so the load-bearing negative assertion it describes never runs and the "public/**" shade exclusions can regress silently. Wire it into scripts/release/validate_staged_release.sh next to validate_source_copyright.sh (and/or a bot.yml step) so it actually executes against built bundle jars.

voonhous replied:
Wired it into the three `validate-bundles*` jobs in `bot.yml`, right after **Build Project** -- that is where built jars actually exist (`packaging/*/target/`, the script's default search dir). The scala-2.12 legs do a full build, so the positive server-bundle assertion runs there too.

Skipped `validate_staged_release.sh`: it only checks the source tarball (checksums, GPG, RAT) and never builds jars, so the check would be a permanent no-op there.

wombatu-kun commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/handlers/TimelineHandler.java:
```diff
@@ -46,4 +135,185 @@ public List<InstantDTO> getLastInstant(String basePath) {
+        String schemaStr = commitMetadata.getMetadata(HoodieCommitMetadata.SCHEMA_KEY);
+        if (schemaStr != null && !schemaStr.isEmpty() && !schemaStr.equals(previousSchema)) {
+          Map<String, String> entry = new LinkedHashMap<>();
+          entry.put("type", previousSchema == null ? "baseline" : "change");
```
When the scan window is truncated (limit smaller than the commit count) the first scanned instant is labeled type="baseline" though it is only the oldest in the window, not the table's original schema. Consider deriving the label from window.truncated so a mid-history schema is not presented as the baseline; minor follow-up.

voonhous replied:
Done, but in the UI rather than the handler: RFC-94 pins the wire contract (`baseline` = "the schema as of the oldest commit scanned", with `window.truncated` as the separate signal), so changing the JSON `type` would diverge from the merged spec. The badge label is now derived from `window.truncated` -- when the window is truncated the oldest entry renders as `oldest scanned` instead of `baseline`.

wombatu-kun commented:
The failing `test-flink-1 (flink2.1)` shard is unrelated to this timeline-UI change. It is a known flaky teardown race in the Flink CDC streaming-read ITs (`ITTestHoodieDataSource.testStreamReadFromSpecifiedCommitWithChangelog`): on job shutdown the split-fetcher thread closes the CDC iterator while the task thread is still draining a batch, surfacing as `NullPointerException at CdcIterators$BaseImageIterator.hasNext` with restarts disabled (`maxNumberRestartAttempts=0`).

A fix is already in flight in #19202 (makes the CDC read iterator teardown thread-safe and tolerates the race in the IT). Re-running the `test-flink-1 (flink2.1)` job should get this PR green in the meantime, since the race usually does not recur.

@voonhous if you have a moment, a review on #19202 would help land the permanent fix.

voonhous committed:
fix(timeline-ui): [HUDI-9315] Keep millisecond precision in parsed instant timestamps

voonhous committed:
fix(timeline-ui): [HUDI-9315] Fix complex-type comparison and rendering in internal-schema diffs

voonhous committed:
fix(timeline-ui): [HUDI-9315] Drop stale clean-range fetch responses after selection changes

voonhous committed:
fix(timeline-ui): [HUDI-9315] Wire bundle UI asset validation into CI bundle jobs

voonhous committed:
fix(timeline-ui): [HUDI-9315] Label truncated-window oldest schema entry as 'oldest scanned', not baseline

wombatu-kun commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/TimelineService.java:
```diff
@@ -230,17 +238,35 @@ private void createApp() {
     ScheduledExecutorScheduler scheduler = new ScheduledExecutorScheduler("TimelineService-JettyScheduler", true, 8);
     server.addBean(scheduler);
 
+    if (timelineServerConf.enableUi && TimelineService.class.getResource("/public/index.html") == null) {
```
This check runs inside createApp(), which startServiceOnPort retries 16x while swallowing every exception, so it never fails fast and the actionable message is lost behind a generic port IOException. Move it ahead of the retry loop, for example into startService.

voonhous replied:
Fixed -- moved the check to the top of `startServiceOnPort`, next to the existing port-range precondition, so it throws once with the actionable message instead of being swallowed by the retry loop. `createApp()` no longer has the guard.

wombatu-kun commented on hudi-timeline-service/src/main/resources/public/js/timeline.js:
```diff
@@ -0,0 +1,1262 @@
+        renderjson.set_show_to_level(1);
+        renderjson.set_icons('\u25B6', '\u25BC');
+        renderjson.set_sort_objects(true);
+        detailBody.appendChild(renderjson(json));
```
This fetch writes detailBody unconditionally, so fast selection changes can paint an older instant's JSON under the current header. Add the same getSelection()[0] !== item.id guard the nested clean-range fetch already uses; follow-up.

voonhous replied:
Fixed -- same `timeline.getSelection()[0] !== item.id` guard added to the details fetch, in both the success and the error handler (a stale failure could equally paint its alert under the current header). Stale responses also no longer trigger the nested clean-range fetch at all.

wombatu-kun commented on hudi-timeline-service/src/test/java/org/apache/hudi/timeline/service/TestUiApi.java:
```diff
@@ -0,0 +1,713 @@
+  // ---------------------------------------------------------------------------
+
+  @Test
+  void testGetInstantDetailsCommitRoundTrip() throws Exception {
```
The savepoint arm of getInstantDetails (the only one that reads the inflight directly, without requestedTwin) has no test. Add a completed-savepoint detail case; follow-up.

voonhous replied:
Added `testGetInstantDetailsCompletedSavepoint` -- writes an inflight + completed savepoint via `HoodieTestTable.addSavepointCommit` and asserts the avro `HoodieSavepointMetadata` fields (`savepointedBy`, `partitionMetadata.*.savepointDataFile`) come back through the Map conversion.

wombatu-kun commented on scripts/release/validate_bundle_ui_assets.sh:
```diff
@@ -0,0 +1,154 @@
+  'hudi-flink*-bundle-*.jar'
+  'hudi-utilities-bundle_*.jar'
+  'hudi-kafka-connect-bundle-*.jar'
+  'hudi-integ-test-bundle-*.jar'
```
hudi-integ-test-bundle is never built in the validate-bundles jobs that run this script, so this pattern always SKIPs and its public/ exclusion is never asserted. Run the script in the integration-tests job after its build, or drop this pattern.

voonhous replied:
It is actually built in those jobs: the scala-2.12 legs run the full reactor with `-DdeployArtifacts=true`, which activates the `integration-tests` profile in the root pom, and that profile adds `packaging/hudi-integ-test-bundle` as a module. Verified in a recent master run of `validate-bundles (scala-2.12, flink1.18, spark3.3)`: the log shows `Building hudi-integ-test-bundle 1.3.0-SNAPSHOT [75/75]`. So the pattern asserts (not SKIPs) in the two scala-2.12 legs; the SKIP path only applies to the subset-build legs, same as for the flink/kafka-connect patterns there.

wombatu-kun commented on scripts/release/validate_bundle_ui_assets.sh:
```diff
@@ -0,0 +1,154 @@
+      echo "Neither 'jar' nor 'unzip' is available to inspect jars [ERROR]"
+      exit 1
+    fi
+    offending=$(echo "$entries" | grep '^public/')
```
An empty jar listing leaves this grep empty, so a truncated or unreadable jar passes the "no public/" check vacuously. Assert the listing contains a known entry like META-INF/ before the OK verdict; follow-up.

voonhous replied:
Fixed -- the negative loop now requires a `META-INF/` entry in the listing before the OK verdict; an empty or unreadable jar counts as a violation. Verified both ways: a truncated fake bundle jar now fails the run, real jars still pass.

voonhous committed:
fix(timeline-ui): [HUDI-9315] Move UI-assets fail-fast check ahead of the port retry loop

voonhous committed:
fix(timeline-ui): [HUDI-9315] Guard instant-details rendering against stale fetch responses

voonhous committed:
test(timeline-ui): [HUDI-9315] Cover the completed-savepoint arm of getInstantDetails

voonhous committed:
fix(timeline-ui): [HUDI-9315] Fail bundle UI asset check on empty or unreadable jar listings
